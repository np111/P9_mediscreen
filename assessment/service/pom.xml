<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>com.mediscreen</groupId>
    <artifactId>mediscreen-common-service-parent</artifactId>
    <version>1.0-SNAPSHOT</version>
    <relativePath>../../common/service-parent/pom.xml</relativePath>
  </parent>
  <artifactId>mediscreen-assessment-service</artifactId>
  <packaging>jar</packaging>

  <dependencies>
    <!-- Local modules -->
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-common-restclient</artifactId>
    </dependency>
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-patients-api</artifactId>
    </dependency>
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-notes-api</artifactId>
    </dependency>
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-assessment-api</artifactId>
    </dependency>
  </dependencies>
</project>
