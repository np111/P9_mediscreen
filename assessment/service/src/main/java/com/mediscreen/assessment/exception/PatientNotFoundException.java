package com.mediscreen.assessment.exception;

import com.mediscreen.common.util.exception.FastRuntimeException;
import java.util.UUID;
import lombok.Getter;

@Getter
public class PatientNotFoundException extends FastRuntimeException {
    private final UUID id;

    public PatientNotFoundException(UUID id) {
        super("Patient not found: " + id);
        this.id = id;
    }
}
