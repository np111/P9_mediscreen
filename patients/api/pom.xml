<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>com.mediscreen</groupId>
    <artifactId>mediscreen</artifactId>
    <version>1.0-SNAPSHOT</version>
    <relativePath>../../pom.xml</relativePath>
  </parent>
  <artifactId>mediscreen-patients-api</artifactId>
  <packaging>jar</packaging>

  <dependencies>
    <!-- Local modules -->
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-common-api</artifactId>
    </dependency>

    <!-- Validation API (JSR 380) -->
    <dependency>
      <groupId>javax.validation</groupId>
      <artifactId>validation-api</artifactId>
      <scope>provided</scope>
    </dependency>
  </dependencies>
</project>
