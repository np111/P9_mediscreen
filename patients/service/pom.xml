<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>com.mediscreen</groupId>
    <artifactId>mediscreen-common-service-parent</artifactId>
    <version>1.0-SNAPSHOT</version>
    <relativePath>../../common/service-parent/pom.xml</relativePath>
  </parent>
  <artifactId>mediscreen-patients-service</artifactId>
  <packaging>jar</packaging>

  <dependencies>
    <!-- Local modules -->
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-patients-api</artifactId>
    </dependency>

    <!-- Spring Boot -->
    <!-- * Starter Data JPA (w/ MariaDB) -->
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-data-jpa</artifactId>
    </dependency>
    <dependency>
      <groupId>org.mariadb.jdbc</groupId>
      <artifactId>mariadb-java-client</artifactId>
      <scope>runtime</scope>
    </dependency>

    <!-- Flyway -->
    <dependency>
      <groupId>org.flywaydb</groupId>
      <artifactId>flyway-core</artifactId>
    </dependency>

    <!-- Google libphonenumber -->
    <dependency>
      <groupId>com.googlecode.libphonenumber</groupId>
      <artifactId>libphonenumber</artifactId>
    </dependency>
  </dependencies>
</project>
