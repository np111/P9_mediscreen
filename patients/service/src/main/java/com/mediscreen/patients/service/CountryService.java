package com.mediscreen.patients.service;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mediscreen.patients.api.model.Countries;
import com.mediscreen.patients.api.model.Country;
import com.mediscreen.patients.util.LocalizedMapUtil;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.Data;
import lombok.Getter;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.stereotype.Service;

@Service
public class CountryService implements InitializingBean {
    private Set<String> countriesCodes;
    private @Getter Set<String> supportedLocales;
    private Countries countriesFallback;
    private Map<String, Countries> countriesByLocale;

    @Override
    public void afterPropertiesSet() throws Exception {
        List<Iso3166Country> iso3166Countries;
        try (Reader rd = new InputStreamReader(CountryService.class.getResourceAsStream("/iso3166.json"), StandardCharsets.UTF_8)) {
            iso3166Countries = new ObjectMapper()
                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                    .readerForListOf(Iso3166Country.class)
                    .readValue(rd);
        }

        Set<String> countriesCodes = iso3166Countries.stream().map(Iso3166Country::getAlpha2).collect(Collectors.toSet());

        Set<String> supportedLocales = iso3166Countries.stream().flatMap(country -> country.getName().keySet().stream()).collect(Collectors.toSet());

        Map<String, Countries> countriesByLocale = new HashMap<>();
        for (String locale : supportedLocales) {
            Countries.Builder countries = Countries.builder().locale(locale);
            iso3166Countries.forEach(country -> {
                String code = country.getAlpha2();
                String name = country.getName().getOrDefault(locale, country.getName().getOrDefault("en", code));
                countries.country(code, Country.builder().code(code).name(name).build());
            });
            countriesByLocale.put(locale, countries.build());
        }

        this.countriesCodes = Collections.unmodifiableSet(countriesCodes);
        this.supportedLocales = Collections.unmodifiableSet(supportedLocales);
        this.countriesFallback = Objects.requireNonNull(countriesByLocale.get("en"));
        this.countriesByLocale = Collections.unmodifiableMap(countriesByLocale);
    }

    public boolean isCountryCode(String value) {
        return countriesCodes.contains(value);
    }

    public Countries getLocalizedCountries(String locale) {
        return LocalizedMapUtil.get(countriesByLocale, locale, countriesFallback);
    }

    @Data
    private static class Iso3166Country {
        private String alpha2;
        private Map<String, String> name;
    }
}
