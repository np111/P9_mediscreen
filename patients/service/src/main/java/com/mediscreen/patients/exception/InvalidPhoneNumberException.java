package com.mediscreen.patients.exception;

import com.mediscreen.common.util.exception.FastRuntimeException;
import lombok.Getter;

@Getter
public class InvalidPhoneNumberException extends FastRuntimeException {
    private final int countryCallingCode;
    private final String number;
    private final String parseError;

    public InvalidPhoneNumberException(int countryCallingCode, String number, String parseError, String message) {
        super(message);
        this.countryCallingCode = countryCallingCode;
        this.number = number;
        this.parseError = parseError;
    }
}
