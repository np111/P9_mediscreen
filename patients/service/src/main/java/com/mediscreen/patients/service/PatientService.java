package com.mediscreen.patients.service;

import com.mediscreen.patients.api.model.Patient;
import com.mediscreen.patients.exception.PatientNotFoundException;
import java.util.List;
import java.util.UUID;

public interface PatientService {
    List<Patient> list();

    Patient get(UUID id) throws PatientNotFoundException;

    Patient create(Patient patient);

    Patient update(UUID id, Patient patient) throws PatientNotFoundException;

    void delete(UUID id) throws PatientNotFoundException;
}
