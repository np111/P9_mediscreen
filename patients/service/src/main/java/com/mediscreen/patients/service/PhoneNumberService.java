package com.mediscreen.patients.service;

import com.google.i18n.phonenumbers.NumberParseException;
import com.google.i18n.phonenumbers.PhoneNumberUtil;
import com.google.i18n.phonenumbers.Phonenumber.PhoneNumber;
import com.mediscreen.patients.api.model.Country;
import com.mediscreen.patients.api.model.PhoneCountries;
import com.mediscreen.patients.api.model.PhoneCountry;
import com.mediscreen.patients.exception.InvalidPhoneNumberException;
import com.mediscreen.patients.util.LocalizedMapUtil;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Service
public class PhoneNumberService implements InitializingBean {
    private static final Pattern PHONE_NUMBER_PATTERN = Pattern.compile("^[1-9][0-9]{0,3} [1-9][0-9]*$");

    private final PhoneNumberUtil pnu = PhoneNumberUtil.getInstance();
    private final CountryService countryService;

    private PhoneCountries phoneCountriesFallback;
    private Map<String, PhoneCountries> phoneCountriesByLocale;

    @Override
    public void afterPropertiesSet() throws Exception {
        Map<String, PhoneCountries> phoneCountriesByLocale = new HashMap<>();
        for (String locale : countryService.getSupportedLocales()) {
            phoneCountriesByLocale.put(locale, PhoneCountries.builder()
                    .locale(locale)
                    .countries(pnu.getSupportedRegions().stream()
                            .map(region -> PhoneCountry.builder()
                                    .regionCode(region)
                                    .callingCode(pnu.getCountryCodeForRegion(region))
                                    .name(getPhoneCountryName(locale, region))
                                    .build())
                            .sorted(Comparator.comparing(PhoneCountry::getName))
                            .collect(Collectors.toList()))
                    .build());
        }

        this.phoneCountriesFallback = Objects.requireNonNull(phoneCountriesByLocale.get("en"));
        this.phoneCountriesByLocale = Collections.unmodifiableMap(phoneCountriesByLocale);
    }

    private String getPhoneCountryName(String locale, String regionCode) {
        switch (regionCode) {
            case "SH":
                return "Saint Helena";
            case "AC":
                return "Ascension Island";
            case "TA":
                return "Tristan da Cunha";
            case "XK":
                return "Kosovo";
            default:
                Map<String, Country> countries = countryService.getLocalizedCountries(locale).getCountries();
                Country country = countries.get(regionCode.toLowerCase(Locale.ROOT));
                if (country == null) {
                    throw new IllegalArgumentException("Unknown phone country name: " + regionCode);
                }
                return country.getName();
        }
    }

    public PhoneCountries getLocalizedPhoneCountries(String locale) {
        return LocalizedMapUtil.get(phoneCountriesByLocale, locale, phoneCountriesFallback);
    }

    public String parsePhoneNumber(int countryCallingCode, String number) {
        String region = pnu.getRegionCodeForCountryCode(countryCallingCode);
        if ("ZZ".equals(region)) {
            throw new InvalidPhoneNumberException(countryCallingCode, number, "INVALID_COUNTRY_CODE", "Invalid country calling code.");
        }

        PhoneNumber phoneNumber;
        try {
            phoneNumber = pnu.parse(number, region);
        } catch (NumberParseException e) {
            throw new InvalidPhoneNumberException(countryCallingCode, number, e.getErrorType().name(), e.getMessage());
        }

        String ret = phoneNumber.getCountryCode() + " " + phoneNumber.getNationalNumber();
        if (!isValidPhoneNumber(ret)) {
            throw new InvalidPhoneNumberException(countryCallingCode, number, "NOT_A_NUMBER", "The string supplied did not seem to be a phone number.");
        }

        return ret;
    }

    public boolean isValidPhoneNumber(String str) {
        if (!PHONE_NUMBER_PATTERN.matcher(str).matches()) {
            return false;
        }
        int spaceIndex = str.indexOf(' ');
        PhoneNumber phoneNumber = new PhoneNumber();
        phoneNumber.setCountryCode(Integer.parseInt(str.substring(0, spaceIndex)));
        phoneNumber.setNationalNumber(Long.parseLong(str.substring(spaceIndex + 1)));
        return pnu.isValidNumber(phoneNumber);
    }
}
