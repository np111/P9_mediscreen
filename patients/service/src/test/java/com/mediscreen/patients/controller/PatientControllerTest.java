package com.mediscreen.patients.controller;

import com.mediscreen.patients.exception.PatientNotFoundException;
import com.mediscreen.patients.mock.PatientMockData;
import com.mediscreen.patients.mock.TestConfig;
import com.mediscreen.patients.service.PatientService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.web.servlet.MockMvc;

import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doThrow;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;

@WebMvcTest(PatientController.class)
@Import({TestConfig.class})
class PatientControllerTest extends AbstractControllerTest {
    @MockBean
    private PatientService patientService;

    @Autowired
    private PatientMockData patientMockData;

    @Autowired
    private MockMvc mockMvc;

    @Test
    void listPatients() throws Exception {
        doAnswer(m -> patientMockData.getPatientModels())
                .when(patientService).list();

        toMatchSnapshot(
                mockMvc.perform(get("/patient")).andReturn()
        );
    }

    @Test
    void getPatient() throws Exception {
        doAnswer(m -> patientMockData.getPatientModelA())
                .when(patientService).get(patientMockData.getPatientIdA());
        doThrow(new PatientNotFoundException(patientMockData.getPatientIdB()))
                .when(patientService).get(patientMockData.getPatientIdB());

        toMatchSnapshot(
                mockMvc.perform(get("/patient/" + patientMockData.getPatientIdA())).andReturn(),
                mockMvc.perform(get("/patient/" + patientMockData.getPatientIdB())).andReturn()
        );
    }

    @Test
    void createPatient() {
        // TBD
    }

    @Test
    void updatePatient() {
        // TBD
    }

    @Test
    void deletePatient() throws Exception {
        doAnswer(m -> patientMockData.getPatientModelA())
                .when(patientService).delete(patientMockData.getPatientIdA());
        doThrow(new PatientNotFoundException(patientMockData.getPatientIdB()))
                .when(patientService).delete(patientMockData.getPatientIdB());

        toMatchSnapshot(
                mockMvc.perform(delete("/patient/" + patientMockData.getPatientIdA())).andReturn(),
                mockMvc.perform(delete("/patient/" + patientMockData.getPatientIdB())).andReturn()
        );
    }
}