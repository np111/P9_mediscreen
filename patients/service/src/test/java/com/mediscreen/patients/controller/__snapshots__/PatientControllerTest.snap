com.mediscreen.patients.controller.PatientControllerTest.deletePatient=[
-
> DELETE /patient/b650a2bf-08e1-43ea-965d-f0b87ac21bfb
< HTTP 204
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
""
-
> DELETE /patient/23f40323-d9b7-4b75-bad7-670ed1187d5a
< HTTP 404
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"{\"type\":\"SERVICE\",\"code\":\"PATIENT_NOT_FOUND\",\"message\":\"Patient does not exists\",\"metadata\":{\"id\":\"23f40323-d9b7-4b75-bad7-670ed1187d5a\"}}"
]


com.mediscreen.patients.controller.PatientControllerTest.getPatient=[
-
> GET /patient/b650a2bf-08e1-43ea-965d-f0b87ac21bfb
< HTTP 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"{\"id\":\"b650a2bf-08e1-43ea-965d-f0b87ac21bfb\",\"firstName\":\"John\",\"lastName\":\"Smith\"}"
-
> GET /patient/23f40323-d9b7-4b75-bad7-670ed1187d5a
< HTTP 404
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"{\"type\":\"SERVICE\",\"code\":\"PATIENT_NOT_FOUND\",\"message\":\"Patient does not exists\",\"metadata\":{\"id\":\"23f40323-d9b7-4b75-bad7-670ed1187d5a\"}}"
]


com.mediscreen.patients.controller.PatientControllerTest.listPatients=[
-
> GET /patient
< HTTP 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"[{\"id\":\"b650a2bf-08e1-43ea-965d-f0b87ac21bfb\",\"firstName\":\"John\",\"lastName\":\"Smith\"},{\"id\":\"23f40323-d9b7-4b75-bad7-670ed1187d5a\",\"firstName\":\"Mary\",\"lastName\":\"Johnson\"}]"
]